C7/UIElements/RightClickMenu.cs

```diff
@@ -100,45 +101,94 @@ public override void _Input(InputEvent @event)
 		// say that events reach children before their parents, but the catch is that there are three phases of input processing. The "input"
 		// phase, this function, then "gui input", and finally "unhandled input". If a control eats an event during the "input" phase it won't
 		// proceed to the "gui input" phase where buttons actually respond to it.)
-		} else if (! ((@event is InputEventMouse) && mouseOverMenu))
+		} else if (!((@event is InputEventMouse) && mouseOverMenu)) {
 			this.AcceptEvent();
+		}
 	}
 }
 
 public class RightClickTileMenu : RightClickMenu
 {
-	public RightClickTileMenu(Game game, Tile tile) : base(game)
-	{
+	public RightClickTileMenu(Game game, Tile tile) : base(game) {
 		ResetItems(tile);
 	}
 
-	// assumeUnfortified is a unit that will be considered unfortified regardless of what its isFortified field holds. It's used to work around
-	// the fact that unfortification is done asynchronously through a message to the engine.
-	public void ResetItems(Tile tile, MapUnit assumeUnfortified = null)
-	{
+	// uiUpdatedUnitStates maps unit guid to a boolean that is true if they were fortified
+	// and false if they were selected in the previous action. This is to update the UI
+	// since the actions update the engine asynchronously and otherwise the UI may not
+	// reflect these changes immediately.
+	public void ResetItems(Tile tile, Dictionary<string, bool> uiUpdatedUnitStates = null) {
 		RemoveAll();
+
+		bool isUnitFortified(MapUnit unit) {
+			if (uiUpdatedUnitStates is null || !uiUpdatedUnitStates.ContainsKey(unit.guid)) {
+				return unit.isFortified;
+			}
+			return uiUpdatedUnitStates[unit.guid];
+		}
+
+		string getUnitAction(MapUnit unit, bool isFortified) {
+			if (unit.owner == game.controller) {
+				return isFortified ? "Wake" : "Activate";
+			}
+			return "Contact";
+		}
+
+		int fortifiedCount = 0;
 		foreach (MapUnit unit in tile.unitsOnTile) {
-			bool isFortified = (unit != assumeUnfortified) && unit.isFortified;
-			string action = (unit.owner == game.controller) ?
-				(isFortified ? "Wake" : "Activate") :
-				"Contact";
+
+			bool isFortified = isUnitFortified(unit);
+			fortifiedCount += isFortified ? 1 : 0;
+			string action = getUnitAction(unit, isFortified);
+
 			AddItem($"{action} {unit.Describe()}").Connect("pressed", this, "SelectUnit", new Godot.Collections.Array() {unit.guid});
 		}
+		int unfortifiedCount = tile.unitsOnTile.Count - fortifiedCount;
+
+		if (fortifiedCount > 0) {
+			AddItem($"Wake All ({fortifiedCount} units)").Connect("pressed", this, "ForAll", new Godot.Collections.Array() {tile.xCoordinate, tile.yCoordinate, false});
```

Comment:
instead of using tile.XCoordinate and tile.YCoordinate pair, it might be nice to use 1-dimensional coordinates for tile ID

Reply:
Having gone the 1D route in parts of my scenario editor, I prefer the 2D approach.  The downside is the conversion (calls to tileAt).  But it's easier for humans to reason about the 2D grid, even with it being isometric.  If we need to debug something, X/Y makes it much easier to map up to what's visually happening on the map.  I never want to have to say, "the 2347th tile from the start".  Much easier to say, "The tile at 47, 31".

That said, I believe you could pass the Tile object instead of the x, y coordinates here.  It's all on the front-end, so there's no message about a tile to the engine (only a message about a unit).  It appears that this method converts the Tile to coordinates, and ForAll then converts it back to a Tile on line 173... so passing the Tile would save a conversion both ways.

Reply:
I actually tried using a tile initially, but in order for Godot to pass params to the connected function, the parameters need to be a serializable type, which apparently tile is not, so it seems easier to pass the coordinates than to figure out how to make Tile serializable given it would only saves 1 parameter

Reply:
Oh yeah, I ran into that back in November, forgot about the serializeable requirement for Godot events.  There's a work-around in the form of ParameterWrapper.cs, in the top level of the C7 project, that we use in a few places.

Although given that we only really need the coordinates here, I don't feel compelled to ParameterWrap the Tile.